test/e2e/antreaipam_test.go

```diff
@@ -156,29 +200,215 @@ func testAntreaIPAMPodConnectivityDifferentNodes(t *testing.T, data *TestData) {
+	}
+	startIPString := subnetIPv4RangesMap["0"].Spec.IPRanges[0].Start
+	offset := 2
+	if dedicatedIPPoolKey != nil {
+		startIPString = subnetIPv4RangesMap[*dedicatedIPPoolKey].Spec.IPRanges[0].Start
+	}
+	expectedPodIP := utilnet.AddIPOffset(utilnet.BigForIP(net.ParseIP(startIPString)), offset)
+	assert.True(t, isBelongTo)
+	assert.True(t, reflect.DeepEqual(ipAddressState, &crdv1alpha2.IPAddressState{
+		IPAddress: expectedPodIP.String(),
+		Phase:     crdv1alpha2.IPAddressPhaseAllocated,
+		Owner: crdv1alpha2.IPAddressOwner{
+			Pod: &crdv1alpha2.PodOwner{
+				Name:        podName,
+				Namespace:   testAntreaIPAMNamespace,
+				ContainerID: ipAddressState.Owner.Pod.ContainerID,
+			},
+		},
+	}))
+
+	ipOffsets = []int32{0, 1, 3}
+	size = len(ipOffsets)
+	reservedIPOffsets = ipOffsets
+	_, err = data.updateStatefulSetSize(stsName, testAntreaIPAMNamespace, int32(size))
+	if err != nil {
+		t.Fatalf("Error when updating StatefulSet '%s': %v", stsName, err)
+	}
+	if err := data.waitForStatefulSetPods(defaultTimeout, stsName, testAntreaIPAMNamespace); err != nil {
+		t.Fatalf("Error when waiting for StatefulSet Pods to get IPs: %v", err)
+	}
+	checkStatefulSetIPPoolAllocation(t, data, stsName, testAntreaIPAMNamespace, ipPoolName, ipOffsets, reservedIPOffsets)
+
+	data.deletePodAndWait(defaultTimeout, testAntreaIPAMNamespace, podName)
+	_, err = data.restartStatefulSet(stsName, testAntreaIPAMNamespace)
+	if err != nil {
+		t.Fatalf("Error when restarting StatefulSet '%s': %v", stsName, err)
+	}
+	time.Sleep(time.Second)
+	if err := data.waitForStatefulSetPods(defaultTimeout, stsName, testAntreaIPAMNamespace); err != nil {
+		t.Fatalf("Error when waiting for StatefulSet Pods to get IPs: %v", err)
+	}
+	checkStatefulSetIPPoolAllocation(t, data, stsName, testAntreaIPAMNamespace, ipPoolName, ipOffsets, reservedIPOffsets)
+
+	cleanup()
+	checkStatefulSetIPPoolAllocation(t, data, stsName, testAntreaIPAMNamespace, ipPoolName, nil, nil)
+}
+
+func checkStatefulSetIPPoolAllocation(tb testing.TB, data *TestData, name string, namespace string, ipPoolName string, ipOffsets, reservedIPOffsets []int32) {
+	ipPool, err := data.crdClient.CrdV1alpha2().IPPools().Get(context.TODO(), ipPoolName, metav1.GetOptions{})
+	if err != nil {
+		tb.Fatalf("Failed to get IPPool %s, err: %+v", ipPoolName, err)
+	}
+	startIP := net.ParseIP(ipPool.Spec.IPRanges[0].Start)
+	expectedIPAddressMap := map[string]*crdv1alpha2.IPAddressState{}
+	for i, offset := range ipOffsets {
+		ipString := utilnet.AddIPOffset(utilnet.BigForIP(startIP), int(offset)).String()
+		podName := fmt.Sprintf("%s-%d", name, i)
+		expectedIPAddressMap[ipString] = &crdv1alpha2.IPAddressState{
+			IPAddress: ipString,
+			Phase:     crdv1alpha2.IPAddressPhaseAllocated,
+			Owner: crdv1alpha2.IPAddressOwner{
+				Pod: &crdv1alpha2.PodOwner{
+					Name:        podName,
+					Namespace:   namespace,
+					ContainerID: "",
+				},
+			},
+		}
+	}
+	for i, offset := range reservedIPOffsets {
```

Comment:
I feel that since `ipOffsets`  is always a subset of `reservedIPOffsets`, this loop above can be merged into this loop?
I don't insist but it would save some lines of code.

Reply:
I think it still need many lines of code to identify if a element in `reservedIPOffsets` is also in `ipOffsets`, but we can try to refine this part in next release.

Reply:
Sounds good